### ISSUE TITLE
May I ask why you built a custom trainer instead of using SFTTrainer?

### ISSUE THREAD

**Issue by CynthiaChuang** [2U1/Gemma3-Finetune#1]
Hi, I was reading the Google AI developers' implementation and noticed that they use Hugging Face's TRL `SFTTrainer` for fine-tuning. I saw that in your project, you’ve developed a custom `Gemma3Trainer` instead.

I’m really curious—what motivated you to build your own trainer? Was there a specific limitation or issue you encountered with existing tools like SFTTrainer that you wanted to address?

Thanks in advance, and I really appreciate the work you’ve shared!

**Comment by 2U1** [comment]
I was inhertinig the Huggingface Trainer so, it's not quite different from using TRL SFTTrainer.
You can set different learninig_rate to each modules with this way and I've modified to save the lora weight when saving checkpoint.
You could use SFTTrainer if you want but, need a little change in the train.py.

**Comment by CynthiaChuang** [comment]
Thanks so much for the reply!

I’ll start by trying out your version for training, and if needed, I might experiment with `SFTTrainer` later on.

By the way, I noticed something in `data.py`—on line 9 there’s this line:

`from qwen_vl_utils import process_vision_info`

It doesn’t seem to be used anywhere, and causes an error when running inside a custom-built Docker image. Just a heads-up—it might be a good idea to remove that import.

Thanks again for sharing your work!

**Comment by 2U1** [comment]
Oh yes thanks I've copy pasted the code from qwen2-vl finetuning and just modified the other logics for gemma3, so I forgot to remove ths import part.
Thank you for letting me know.

**Comment by 2U1** [comment]
@CynthiaChuang When distributed training, all of the batches in the gpu has to pass though a same activation path/forward path (Has to have a same compiled graph).
When gpu0 has only text and gpu 1 has text+image, then they would have a different graph, so it could cause a deadlock when training when gathering the gradients into to main process.
So I've made a dummy and excluded in the calculation of back propagation. Then this would have all same graph.

**Comment by 2U1** [comment]
Well the code is based on llava code so, there is no special reason for it.
Also, basically the loss calculated in the forward function of the model so it dosent differs much. It only calculates token wise metrics additionaly.

**Comment by 2U1** [comment]
For the DPO code, I'll use the DPOTrainer. The code would be updated soon.

**Comment by CynthiaChuang** [comment]
Oh, yeah — I can cross DPOTrainer off my to-do list now. Looking forward to your update! 💖

**Comment by CynthiaChuang** [comment]
Sorry, just wanted to double-check something with you — I noticed that liger-kernel is installed according to the environment.yaml, but it doesn’t seem to be used anywhere in the code?

**Comment by 2U1** [comment]
Oh yes, the environment.yaml was copied from qwen2-vl repo. 
It's not used right now but when liger-kernel is updated it would be used then.

**Comment by 2U1** [comment]
I've updated the code supporting DPO and liger-kernel.
Also there was a little bit of refactoring, but it won't affect much.
